In [1]:
#!pip install numpy==1.21.4
#!pip install pandas==1.3.4
!pip install keras==2.1.6

# Import Libraries

In [2]:
import pandas as pd
import numpy as np

import warnings
warnings.simplefilter('ignore', FutureWarning)

In [3]:
import keras 

Using TensorFlow backend.


In [4]:
#import libraries for models
from keras.models import Sequential

In [5]:
# import layers 
from keras.layers import Dense

In [6]:
#train test split
from sklearn.model_selection import train_test_split

/home/jupyterlab/conda/envs/python/lib/python3.7/site-packages/sklearn/utils/validation.py:37: DeprecationWarning: distutils Version classes are deprecated. Use packaging.version instead.
  LARGE_SPARSE_SUPPORTED = LooseVersion(scipy_version) >= '0.14.0'


In [7]:
#import metrics MSE
from sklearn.metrics import mean_squared_error

# Load data in panda datafrane

In [8]:
df=pd.read_csv('concrete_data.csv')
df.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  Strength  
0            1040.0           676.0   28     79.99  
1            1055.0           676.0   28     61.89  
2             932.0           594.0  270     40.27  
3             932.0           594.0  365     41.05  
4             978.4           825.5  360     44.30

In [9]:
#column names
col_names= df.columns
col_names

Index(['Cement', 'Blast Furnace Slag', 'Fly Ash', 'Water', 'Superplasticizer',
       'Coarse Aggregate', 'Fine Aggregate', 'Age', 'Strength'],
      dtype='object')

# Split in predictors  and target

In [10]:
#target 
target = df['Strength']

In [11]:
#predictors
predictors =df[col_names[col_names != 'Strength']] # all columns except Strength
predictors.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  
0            1040.0           676.0   28  
1            1055.0           676.0   28  
2             932.0           594.0  270  
3             932.0           594.0  365  
4             978.4           825.5  360

# Check if all the data is there

In [12]:
#check if all the data is there
df.isnull().count()

Cement                1030
Blast Furnace Slag    1030
Fly Ash               1030
Water                 1030
Superplasticizer      1030
Coarse Aggregate      1030
Fine Aggregate        1030
Age                   1030
Strength              1030
dtype: int64

# How many columns / features?

In [13]:
df.shape

(1030, 9)

<!-- # instantiate model -->

# compute input shape

In [14]:
#number of column / features
ncols=predictors.shape[1]
ncols

8

# Regression model function

In [15]:
def regression_model():

    # create model
    model = Sequential()
    # input layer with input shape of the pretoctors +  hidden layer with 10 nodes 
    model.add(Dense(10, activation='relu', input_shape=(ncols,)))
    model.add(Dense(1))#output
    
    # compile model usign adam optimaizer 
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model

# B: Normalizing data set(Predictors)



In [17]:
#normalizing predictors
predictors_normalized= ( predictors - predictors.mean() ) / predictors.std()

# function of process 1-3 combined 

# C: Increased to 100 epochs

In [21]:
# the step 1-3
def reg_model_process_normalized():
    #Randomly split the data into a training and test sets by holding 30% of the data for testing.
    X_train_norm, X_test_norm, y_train, y_test = train_test_split(predictors_normalized, target, test_size=0.30, random_state=42)
    #build model
    model=regression_model()  
    #fit with 100 epochs
    model.fit(X_train_norm ,y_train, validation_split=0.3, epochs=100, verbose=2)
    #predict
    y_pred=model.predict(X_test_norm)
    #compute the mean squared error between the predicted concrete strength and the actual concrete strength using the mean_squared_error
    mse = mean_squared_error(y_test, y_pred)
    return mse




In [19]:
mses_normalized=[]
i=0
for i in range(50):
    mses_normalized.append(reg_model_process_normalized())







Train on 504 samples, validate on 217 samples
Epoch 1/100


2024-10-17 20:32:22.705978: I tensorflow/core/platform/cpu_feature_guard.cc:142] Your CPU supports instructions that this TensorFlow binary was not compiled to use: SSE4.1 SSE4.2 AVX AVX2 AVX512F FMA
2024-10-17 20:32:22.711405: I tensorflow/core/platform/profile_utils/cpu_utils.cc:94] CPU Frequency: 2394300000 Hz
2024-10-17 20:32:22.711956: I tensorflow/compiler/xla/service/service.cc:168] XLA service 0x562d58f6e490 executing computations on platform Host. Devices:
2024-10-17 20:32:22.711997: I tensorflow/compiler/xla/service/service.cc:175]   StreamExecutor device (0): <undefined>, <undefined>
2024-10-17 20:32:22.804005: W tensorflow/compiler/jit/mark_for_compilation_pass.cc:1412] (One-time warning): Not using XLA:CPU for cluster because envvar TF_XLA_FLAGS=--tf_xla_cpu_global_jit was not set.  If you want XLA:CPU, either set that envvar, or use experimental_jit_scope to enable XLA:CPU.  To confirm that XLA is active, pass --vmodule=xla_compilation_cache=1 (as a proper command-line fl

 - 0s - loss: 1735.9628 - val_loss: 1616.3855
Epoch 2/100
 - 0s - loss: 1723.6368 - val_loss: 1604.1865
Epoch 3/100
 - 0s - loss: 1711.4394 - val_loss: 1592.8113
Epoch 4/100
 - 0s - loss: 1699.9316 - val_loss: 1581.7797
Epoch 5/100
 - 0s - loss: 1688.8572 - val_loss: 1571.0725
Epoch 6/100
 - 0s - loss: 1678.1943 - val_loss: 1560.6647
Epoch 7/100
 - 0s - loss: 1667.6670 - val_loss: 1550.8827
Epoch 8/100
 - 0s - loss: 1657.6493 - val_loss: 1541.2329
Epoch 9/100
 - 0s - loss: 1647.8708 - val_loss: 1531.9600
Epoch 10/100
 - 0s - loss: 1638.2930 - val_loss: 1522.7630
Epoch 11/100
 - 0s - loss: 1628.8235 - val_loss: 1513.8226
Epoch 12/100
 - 0s - loss: 1619.4896 - val_loss: 1504.9815
Epoch 13/100
 - 0s - loss: 1610.2968 - val_loss: 1496.1261
Epoch 14/100
 - 0s - loss: 1600.9324 - val_loss: 1487.4815
Epoch 15/100
 - 0s - loss: 1591.7031 - val_loss: 1478.8503
Epoch 16/100
 - 0s - loss: 1582.4832 - val_loss: 1470.1593
Epoch 17/100
 - 0s - loss: 1573.1119 - val_loss: 1461.5134
Epoch 18/100
 - 0s


# How does the mean of the mean squared errors compare to that from Step B?

In [22]:
df_mses_normalized = pd.DataFrame(data=mses_normalized, columns=['MSE with Normalized_Predictors'])


In [23]:
# Mean of Normalized predictor MSE - Step B
df_mses_normalized.mean()

MSE with Normalized_Predictors    212.27002
dtype: float64


### Step C MSE with 100 epochs:  212.27002


Step C with 100 Epochs is MSE of 212  which improved three times better than Step B with 50 epochs (MSE of 633)